pgt/src/PGT/Output/Text.hs

```diff
@@ -11,15 +15,53 @@ module PGT.Output.Text
 where
 
 import Data.Attoparsec.Text (Parser)
+import Data.Int (Int)
+import Data.Tuple (fst)
 import Data.Word (Word8)
 import GHC.Integer (Integer)
 import PGT.Prelude
 
-import qualified Data.Attoparsec.Text as Text
-import qualified Data.Char            as Char
-import qualified Data.Foldable        as Foldable
-import qualified Data.Text            as Text
-import qualified GHC.Err              as Err
+import qualified Data.Attoparsec.Internal.Types as Text
+import qualified Data.Attoparsec.Text           as Text
+import qualified Data.Char                      as Char
+import qualified Data.Foldable                  as Foldable
+import qualified Data.Text                      as Text
+import qualified GHC.Err                        as Err
+
+eitherFail :: Either Text a -> Parser a
+eitherFail = either failP pure
+
+eitherImpureError :: Either Text a -> Parser a
+eitherImpureError = either errorP pure
+
+errorP :: Text -> Parser a
+errorP = mkParserFailure Err.error
+
+failP :: Text -> Parser a
+failP = mkParserFailure fail
+
+mkParserFailure
+  :: (String -> Parser a)
+  -> Text
+  -> Parser a
+mkParserFailure onFailure message = do
+  linePosition <- getLinePosition
+
+  onFailure
+    $ convert message <> " at line: " <> show linePosition
+  where
+    getLinePosition :: Parser Int
+    getLinePosition
+      = Text.Parser
+      $ \textBuffer position more _failure success ->
+            success textBuffer position more . getLineNumber position $ showc textBuffer
+      where
+        getLineNumber :: Text.Pos -> Text -> Int
+        getLineNumber position string
+          = Foldable.length
+          . Text.splitOn "\\n"
```

Comment:
Comment: `Text.unlines` doesn't work here possibly because Attoparsec seems to use its own custom Text Buffer types that some turn new line chars into text.